vectordb_orm/backends/pinecone/pinecone.py

```diff
@@ -0,0 +1,256 @@
+
+        pinecone.init(
+            api_key=api_key,
+            environment=environment,
+        )
+
+    def create_collection(self, schema: Type[VectorSchemaBase]):
+        collection_name = self.transform_collection_name(schema.collection_name())
+        current_indexes = pinecone.list_indexes()
+        if collection_name in current_indexes:
+            info("Collection already created...")
+            return
+
+        info("Creating collection, this could take 30s-5min...")
+
+        self._assert_valid_collection_name(collection_name)
+        self._assert_has_primary(schema)
+        self._assert_valid_embedding_field(schema)
+
+        # Pinecone allows for dynamic keys on each object
+        # However we need to pre-provide the keys we want to search on
+        # This does increase memory size so we can consider adding an explict index flag
+        # for the metadata fields that we want to allow search on
+        # https://docs.pinecone.io/docs/manage-indexes#selective-metadata-indexing
+        metadata_config = {
+            "indexed": [
+                key
+                for key in schema.__annotations__.keys()
+            ]
+        }
+
+        _, embedding_field = self._get_embedding_field(schema)
+        index : PineconeIndex = embedding_field.index
+
+        pinecone.create_index(
+            name=collection_name,
+            dimension=embedding_field.dim,
+            metric=index.metric_type.value,
+            metadata_config=metadata_config
+        )
+        return pinecone.Index(index_name=collection_name)
+
+    def clear_collection(self, schema: Type[VectorSchemaBase]):
+        collection_name = self.transform_collection_name(schema.collection_name())
+
+        current_indexes = pinecone.list_indexes()
+        if collection_name not in current_indexes:
+            return
+
+        index = pinecone.Index(index_name=collection_name)
+        delete_response = index.delete(delete_all=True)
+        if delete_response:
+            # Success should have an empty dict
+            raise ValueError(f"Failed to clear collection {collection_name}: {delete_response}")
+
+    def delete_collection(self, schema: Type[VectorSchemaBase]):
+        collection_name = self.transform_collection_name(schema.collection_name())
+    
+        # Pinecone throws a 404 if the index doesn't exist
+        current_indexes = pinecone.list_indexes()
+        if collection_name not in current_indexes:
+            return
+
+        pinecone.delete_index(collection_name)
+
+    def insert(self, entity: VectorSchemaBase) -> list:
+        schema = entity.__class__
+        collection_name = self.transform_collection_name(schema.collection_name())
+
+        schema = entity.__class__
+        embedding_field_key, _ = self._get_embedding_field(schema)
+        primary_key = self._get_primary(schema)
+
+        id = uuid4().int & (1<<64)-1
```

Comment:
🤢

Reply:
maybe clean this up? or add a comment describing the id gen logic and why it's done this way to guarantee uniqueness?

Reply:
i for one like uuid + timestamps but maybe that's just me